crates/astria-merkle/src/lib.rs

```diff
@@ -0,0 +1,718 @@
+    /// let root = tree.root();
+    /// let proof = tree
+    ///     .construct_proof(6)
+    ///     .expect("leaf 7 must be inside the tree");
+    /// let reconstructed_root = proof.reconstruct_root_with_leaf(&[7_u8; 32]);
+    /// assert_eq!(root, reconstructed_root);
+    /// ```
+    #[must_use]
+    pub fn reconstruct_root_with_leaf(&self, node: &[u8]) -> [u8; 32] {
+        self.reconstruct_root_with_leaf_hash(hash_leaf(node))
+    }
+}
+
+/// A low-level API to construct a leaf-hash ad-hoc without needing to allocate it.
+///
+/// [`LeafBuilder`] can only be created by using the [`Tree::build_push_leaf`] method.
+/// The leaf is built and the tree updated once `LeafBuilder` goes out of scope.
+///
+/// No two leaf builders can exist at the same time because a builder holds a mutable
+/// reference to its tree during its lifetime.
+///
+/// See [`Tree::build_leaf`] for usage.
+pub struct LeafBuilder<'a> {
+    tree: &'a mut Tree,
+    hasher: Option<Sha256>,
+}
+
+impl<'a> LeafBuilder<'a> {
+    /// Takes ownership of the builder, dropping it.
+    ///
+    /// This method causes the leaf builder to go out of scope, causing it
+    /// to be dropped and triggering an update of the merkle tree. Use this
+    /// if the builder is explicitly bound.
+    ///
+    /// Usually calling this is not necessary because writes to the builder can
+    /// be chained, but this is still useful if some other operation has to happen
+    /// between writes, requiring the builder to be found.
+    ///
+    /// # Examples
+    /// ```
+    /// let mut tree_1 = astria_merkle::Tree::new();
+    /// let mut builder = tree_1.build_leaf();
+    /// builder.write(b"hello");
+    /// builder.write(b"world");
+    /// builder.finish();
+    ///
+    /// let mut tree_2 = astria_merkle::Tree::new();
+    /// tree_2.build_leaf().write(b"hello").write(b"world");
+    /// assert_eq!(tree_1.root(), tree_2.root());
+    /// ```
+    pub fn finish(self) {}
+
+    /// Writes `bytes` into the builder, appending to the leaf.
+    ///
+    /// See [`Tree::build_leaf`] for example usage.
+    #[allow(clippy::missing_panics_doc)] // invariant of the system
+    pub fn write(&mut self, bytes: &[u8]) -> &mut Self {
+        let hasher = self
+            .hasher
+            .as_mut()
+            .expect("hasher is set during the lifetime of the leaf builder");
+        hasher.update(bytes);
+        self
+    }
+}
+
+impl<'a> Drop for LeafBuilder<'a> {
+    fn drop(&mut self) {
+        let Self {
+            tree,
+            hasher,
+        } = self;
+        let leaf_hash: [u8; 32] = hasher
+            .take()
+            .expect("hasher is set during the leaf builder's lifetime and only taken on drop")
+            .finalize()
+            .into();
+        if tree.nodes.is_empty() {
+            tree.nodes.extend_from_slice(&leaf_hash);
+            return;
+        }
+        // append 2 * 32 = 64 zeros
+        tree.nodes.extend_from_slice(&[0; 64]);
```

Comment:
why 64?

Reply:
isn't it possible that adding the new leaf adds a new level to the tree, thus adding `log2(new_len)` nodes?

Reply:
This comes from memory structure of the tree: For each leaf-node you have to push 1 extra branch node so odd-indexed elements are leaves, even-indexed elements are branches.

Reply:
> isn't it possible that adding the new leaf adds a new level to the tree, thus adding `log2(new_len)` nodes?

No, that does not happen. You can see that if you draw the evolution of the tree:

```
# Single leaf
x

# Two leaves
  x
 /  \
x  x

# Three leaves
      x
    /   \ 
  x     |
 /  \    |
x  x  x
```
Have a look at the module doc comment at the top of `lib.rs` that shows the evolution and indexing of the tree.